sign_verify_file_ssh.sh

```diff
@@ -0,0 +1,128 @@
+        # Convert PEM key to OpenSSH format
+        cp "$key_file" "$output_file" && ssh-keygen -c -C "Converted from PEM" -f "$output_file" || {
+            echo "Error: Failed to convert PEM key to OpenSSH."
+            exit 1
+        }
+
+        # Generate the public key
+        ssh-keygen -y -f "$key_file" > "$output_file.pub" || {
+            echo "Error: Failed to generate public key from PEM key."
+            exit 1
+        }
+    else
+        echo "$key_file doesn't look like a PEM format key!"
+        exit 1
+    fi
+}
+
+# Sign mode
+if [ "$MODE" == "sign" ]; then
+    PRIVATE_KEY_PEM="$3"
+    PRIVATE_KEY="conversion_id"
+    SIG_FILE="${FILE_TO_SIGN}.sig"
+    PUB_KEY="${PRIVATE_KEY}.pub"
+
+    if [ ! -f "$PRIVATE_KEY_PEM" ]; then
+        echo "Error: Private key '$PRIVATE_KEY_PEM' not found."
+        exit 1
+    fi
+
+    # Convert PEM key to OpenSSH if needed
+    echo "Converting SSH key to OpenSSH format..."
+    convert_pem_to_ssh "$PRIVATE_KEY_PEM" "$PRIVATE_KEY"
+
+    # Sign the file
+    echo "Signing the file..."
+    ssh-keygen -Y sign -f "$PRIVATE_KEY" -n file "$FILE_TO_SIGN"
+
+    if [ ! -f "$SIG_FILE" ]; then
+        echo "Error: Signing failed."
+        rm -f "$PRIVATE_KEY" "$PUB_KEY"
+        exit 1
+    fi
+
+    echo "Signature created: $SIG_FILE"
+    echo -e "\nAn allowed signatures file for this key would have contents like:\n"
+    echo -e "some_name $(cat "$PUB_KEY")\n"
+
+    # Verify the signature
+    echo -e "Validating the signature of the file..."
+    if ssh-keygen -Y check-novalidate -n file -f "$PUB_KEY" -s "$SIG_FILE" < "$FILE_TO_SIGN"; then
+        echo "- Signature validation successful."
+    else
+        echo "- Signature validation failed."
+        rm -f "$PRIVATE_KEY" "$PUB_KEY"
+        exit 1
+    fi
+
+    rm -f "$PRIVATE_KEY" "$PUB_KEY"
+
+# Verify mode
+elif [ "$MODE" == "verify" ]; then
+    ALLOWED_SIGNERS_FILE="$3"
+    SIG_FILE="${4:-${FILE_TO_SIGN}.sig}"
+
+    if [ ! -f "$ALLOWED_SIGNERS_FILE" ]; then
+        echo "Error: Allowed signers file '$ALLOWED_SIGNERS_FILE' not found."
+        exit 1
+    fi
+
+    if [ ! -f "$SIG_FILE" ]; then
+        echo "Error: Signature file '$SIG_FILE' not found."
+        exit 1
+    fi
+
+    # Loop through each line of the allowed_signers file
+    while read -r identity pub_key; do
+        # Check if identity and public key are not empty
+        if [[ -n "$identity" && -n "$pub_key" ]]; then
+            echo "Verifying signature for identity: $identity"
+
+            # Use ssh-keygen to verify the signature (example with RSA)
+            if ssh-keygen -Y verify -f "$ALLOWED_SIGNERS_FILE" -n file -I "$identity" -s "$SIG_FILE" < "$FILE_TO_SIGN"; then
+                echo "Signature is valid for identity: $identity"
+                exit 0  # Exit once we find a valid signature
+            else
+                echo "Invalid signature for identity: $identity"
+            fi
+        fi
+    done < "$ALLOWED_SIGNERS_FILE"
```

Comment:
Is this the equivalent to importing GPG public keys?

Reply:
I haven't used GPG keys in 20 years so I'm not sure 😉 

This cycles through a given authorised signers file, checking to see if the tarball has been signed by any of the identities. Looking it up now, it seems you can give a wildcard `'*'` to the `-I` option, so the loop here may not be necessary.

Reply:
Wildcards don't work so the loop remains